File: csharp/TellDontAsk.Tests/UseCase/OrderShipmentUseCaseTest.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TellDontAsk.Domain;
using TellDontAsk.Tests.Doubles;
using TellDontAsk.UseCase;

namespace TellDontAsk.Tests.UseCase
{
    [TestFixture]
    public class OrderShipmentUseCaseTest
    {
        private TestOrderRepository orderRepository;
        private TestShipmentService shipmentService;
        private OrderShipmentUseCase useCase;

        [SetUp]
        public void SetUp()
        {
            orderRepository = new TestOrderRepository();
            shipmentService = new TestShipmentService();
            useCase = new OrderShipmentUseCase(orderRepository, shipmentService);
        }

        [Test]
        public void shipApprovedOrder()
        {
            Order initialOrder = new Order();
            initialOrder.setId(1);
            initialOrder.setStatus(OrderStatus.APPROVED);
            orderRepository.addOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest();
            request.setOrderId(1);

            useCase.run(request);

            Assert.That(orderRepository.getSavedOrder().getStatus(), Is.EqualTo(OrderStatus.SHIPPED));
            Assert.That(shipmentService.getShippedOrder(), Is.EqualTo(initialOrder));
        }

        [Test]
        public void createdOrdersCannotBeShipped()
        {
            Order initialOrder = new Order();
            initialOrder.setId(1);
            initialOrder.setStatus(OrderStatus.CREATED);
            orderRepository.addOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest();
            request.setOrderId(1);

            Assert.That(() => useCase.run(request),
              Throws.TypeOf<OrderCannotBeShippedException>());


            Assert.That(orderRepository.getSavedOrder(), Is.Null);
            Assert.That(shipmentService.getShippedOrder(), Is.Null);
        }

        [Test]
        public void rejectedOrdersCannotBeShipped()
        {
            Order initialOrder = new Order();
            initialOrder.setId(1);
            initialOrder.setStatus(OrderStatus.REJECTED);
            orderRepository.addOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest();
            request.setOrderId(1);

            Assert.That(() => useCase.run(request),
                         Throws.TypeOf<OrderCannotBeShippedException>());
        }

        [Test]
        public void shippedOrdersCannotBeShippedAgain()
        {
            Order initialOrder = new Order();
            initialOrder.setId(1);
            initialOrder.setStatus(OrderStatus.SHIPPED);
            orderRepository.addOrder(initialOrder);

            OrderShipmentRequest request = new OrderShipmentRequest();
            request.setOrderId(1);

            Assert.That(() => useCase.run(request),
                          Throws.TypeOf<OrderCannotBeShippedTwiceException>());
        }
    }
}

File: php/tests/UseCase/OrderShipmentUseCaseTest.php
<?php

namespace Tests\Pitchart\TellDontAskKata\UseCase;

use PHPUnit\Framework\Assert;
use PHPUnit\Framework\TestCase;
use Pitchart\TellDontAskKata\Domain\Order;
use Pitchart\TellDontAskKata\Domain\OrderStatus;
use Pitchart\TellDontAskKata\UseCase\OrderCannotBeShippedException;
use Pitchart\TellDontAskKata\UseCase\OrderCannotBeShippedTwiceException;
use Pitchart\TellDontAskKata\UseCase\OrderShipmentRequest;
use Pitchart\TellDontAskKata\UseCase\OrderShipmentUseCase;
use Tests\Pitchart\TellDontAskKata\Doubles\InMemoryOrderRepository;
use Tests\Pitchart\TellDontAskKata\Doubles\TestShipmentService;

class OrderShipmentUseCaseTest extends TestCase
{
    private InMemoryOrderRepository $orderRepository;

    private OrderShipmentUseCase $useCase;
    private TestShipmentService $shipmentService;

    protected function setUp(): void
    {
        parent::setUp();
        $this->orderRepository = new InMemoryOrderRepository();
        $this->shipmentService = new TestShipmentService();

        $this->useCase = new OrderShipmentUseCase($this->orderRepository, $this->shipmentService);
    }

    public function test_ship_approved_order(): void
    {
        $initialOrder = (new Order)->setStatus(OrderStatus::Approved)->setId(1);

        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderShipmentRequest())->setId(1);

        $this->useCase->run($request);

        $savedOrder = $this->orderRepository->getSavedOrder();
        Assert::assertEquals(OrderStatus::Shipped, $savedOrder->getStatus());
    }


    public function test_created_orders_can_not_be_shipped()
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Created)->setId(1);

        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderShipmentRequest())->setId(1);

        $this->expectException(OrderCannotBeShippedException::class);

        $this->useCase->run($request);

        Assert:: assertNull($this->orderRepository->getSavedOrder());
        Assert:: assertNull($this->shipmentService->getShippedOrder());
    }


    public function test_rejected_orders_can_not_be_shipped()
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Rejected)->setId(1);

        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderShipmentRequest())->setId(1);

        $this->expectException(OrderCannotBeShippedException::class);

        $this->useCase->run($request);

        Assert:: assertNull($this->orderRepository->getSavedOrder());
        Assert:: assertNull($this->shipmentService->getShippedOrder());
    }


    public function test_shipped_orders_can_not_be_shipped_again()
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Shipped)->setId(1);

        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderShipmentRequest())->setId(1);

        $this->expectException(OrderCannotBeShippedTwiceException::class);

        $this->useCase->run($request);

        Assert:: assertNull($this->orderRepository->getSavedOrder());
        Assert:: assertNull($this->shipmentService->getShippedOrder());
    }

}
